### DISCUSSION TITLE
Request for Documentation on Understanding the Code of a Device Service.

### DISCUSSION THREAD

**Discussion by nicobousquet** [edgexfoundry/community#105] [Q&A]
Greetings,

As a student, I hold a deep-seated interest in the domain of IoT and software development at large. The EdgeX project has particularly caught my attention, and I am keen to comprehend the code of a device service to extend my contribution and partake in the community. Nonetheless, I find the task rather daunting due to the dearth of comments in the code. Currently, I am engrossed in studying the various bootstrap handlers invoked by the function RunAndReturnWaitGroup(). I humbly inquire if any of you are privy to a documentation elucidating the code and its architecture (inclusive of interfaces, all the invoked go routines and their respective functions...)?

Respectfully,
Nicolas.

**Comment by farshidtz** [comment]
Please refer to https://docs.edgexfoundry.org/2.3/getting-started/Ch-GettingStartedSDK-Go/

**Reply by nicobousquet** [reply]
Dear Farshid,

I would like to express my gratitude for sharing the link with me. I must admit that I had already come across this resource before, and although it offers a valuable introduction, it lacks the necessary depth to fully comprehend the workings of a device service.

To provide you with more detailed information about my search, I have noticed that, besides the driver component, all device services follow a similar structure that utilizes functions from the device SDK. I am keen to gain insight into this common feature of device services, particularly regarding the bootstrap and handlers sections. Regrettably, I have not been able to locate relevant documentation within the link you provided. It is possible that such documentation exists elsewhere, but I have yet to discover it.

Thank you again for your assistance in this matter.

Sincerely

**Reply by nicobousquet** [reply]
To elaborate further on my previous message, I am currently seeking a comprehensive collection of UML diagrams that accurately represent all the structures employed within a device service, while also explicitly portraying their interrelationships in terms of composition and interface implementation.

**Reply by cloudxxx8** [reply]
We don't have the detailed document for the current soruce code.  We made some several years ago for the initial version, but they are outdated.
We did lots of refactor, and tried to make the code readable.  Once you have the basic Golang knowledge, you should be able to read the code directly.  Please just write down the specific questions you encounter.

**Reply by nicobousquet** [reply]
Thank you for your response. I must admit that I am not fully acquainted with all the features implemented in a device service's code. Broadly, I understand that a device service enables communication with a device through its driver and facilitates communication with the core metadata via REST requests. However, I suspect that there are other functionalities that I am not aware of.

Could you please explain the primary functionalities implemented in a device service and where they are located in the code? For instance, could you clarify where the configuration, profiles, and devices files are parsed? Additionally, I am curious about the role of each bootstrap handler (httpServer.BootstrapHandler, messaging.BootstrapHandler, clients.BootstrapHandler, autoevent.BootstrapHandler, NewBootstrap(router).BootstrapHandler, autodiscovery.BootstrapHandler, handlers.NewStartMessage(serviceName, serviceVersion).BootstrapHandler).

If possible, could you provide an overview of how the DeviceService structure works? I am particularly interested in what is the role of each attribute of the DeviceService structure, such as LoggingClient, RegistryClient, SecretProvider, edgexClients, controller, config, deviceService, driver, discovery, manager, asyncCh, deviceCh, initialized, dic, flags, configProcessor, ctx, and wg. This will allow me to understand better the working of a device service.

Thank you for your time and expertise.
